### PR TITLE
Unpin notebook and tornado

### LINKED ISSUE #6131: Unpin tornado dependency
#6123 rolled back our dependency on `notebook` due to some upstream issues with async kernel starts (or possibly tornado 6). We need to identify and fix these issues before making a release.

Edit: we have resolved the immediate issue with notebook in 5.7.8, but issues with tornado 6 remain.

### PR DESCRIPTION
**Pull request by ian-r-rose**
Attempt to fix #6131.

### CODE REVIEW

**Comment by jasongrout** [comment]
Can you pull in the other changes from #6138, like printing out the installed packages, and removing the tornado pin in the travis installation script?

**Comment by ian-r-rose** [comment]
I should state that I don't think this is a final fix. I don't think that the extra `awaits` I added should be necessary here, and there are some very subtle low-level things going on in tornado 6 that I don't fully understand (perhaps @minrk can weigh in?).

But it seems like the server hits exceptions if there are multiple requests that come in in quick succession to the kernel WS endpoint (possibly, but not necessarily related to https://github.com/tornadoweb/tornado/commit/e69becbf8c994be5bfa6584ef72df594b27934b5?).

@jasongrout I can pull in those commits this evening, but feel free to push to this branch sooner if you want to.

**Comment by jasongrout** [comment]
There was one commit with relevant things on top of your changes, which I pushed here.

**Comment by jasongrout** [comment]
Locally, this seems to work just fine now (after upgrading tornado, notebook, pyzmq, and nbconvert). I'll merge this after tests pass to help us test it more.

**Comment by jasongrout** [comment]
@ian-r-rose - I'm seeing those (apparently unpredictable) "Already reading" errors you talked about above. Which is weird since the tests passed for you above, and I'm pretty sure my pip freeze addition wouldn't have changed things.

I'm restarting the tests. I'm also wondering if it may have to do with using python 3.5 instead of 3.6? Perhaps there is something going on with the `await` calls and python 3.5?

**Comment by ian-r-rose** [comment]
@jasongrout I'm really not sure. I don't think of this as a real fix, so much as an attempt to paper over some issues with the server side of things.

**Comment by jasongrout** [comment]
I rebased on master, since we recently merged in some fixes to tests from @afshin that may be relevant about dialogs.

**Comment by gnestor** [comment]
For reference, here is the error we are seeing in binder:

```
default.js:73 WebSocket connection to 'wss://hub.mybinder.org/user/jupyterlab-jupyterlab-demo-t7vvr5fk/api/kernels…ssion_id=c14db80e-eadf-4ff4-8880-590e6ee840f9&token=CGW2h3cwTCixJAzlcniajQ' failed: Error during WebSocket handshake: Unexpected response code: 502
DefaultKernel._createSocket	@	default.js:73
```

**Comment by afshin** [comment]
`502` as in `HTTP 502 - Bad Gateway`? Hm.

**Comment by jasongrout** [comment]
I think that is different than the error I was trying to track down, that I was seeing locally and in the CI:
https://github.com/jupyterlab/jupyterlab/pull/6138#issuecomment-478286447

**Comment by jasongrout** [comment]
Here's an example of the error we're seeing a lot in the tests for this PR, for example:

```
[E 22:10:04.426 JestApp] Uncaught exception GET /api/kernels/463a05ee-d173-40e7-b3e6-2b57affeed16/channels?session_id=5ca9613f-7b99-45e2-afbe-73cefa0cff9a&token=6bc3ef0aa40cde31e6cb96be0d82a5044b3f1b4a47aafb9f (127.0.0.1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='GET', uri='/api/kernels/463a05ee-d173-40e7-b3e6-2b57affeed16/channels?session_id=5ca9613f-7b99-45e2-afbe-73cefa0cff9a&token=6bc3ef0aa40cde31e6cb96be0d82a5044b3f1b4a47aafb9f', version='HTTP/1.1', remote_ip='127.0.0.1')
    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/web.py", line 1699, in _execute
        result = await result
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/gen.py", line 736, in run
        yielded = self.gen.throw(*exc_info)  # type: ignore
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/notebook/services/kernels/handlers.py", line 241, in get
        yield super(ZMQChannelsHandler, self).get(kernel_id=kernel_id)
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/gen.py", line 729, in run
        value = future.result()
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/gen.py", line 736, in run
        yielded = self.gen.throw(*exc_info)  # type: ignore
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/notebook/base/zmqhandlers.py", line 297, in get
        yield gen.maybe_future(res)
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/gen.py", line 729, in run
        value = future.result()
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/websocket.py", line 278, in get
        await self.ws_connection.accept_connection(self)
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/websocket.py", line 881, in accept_connection
        await self._accept_connection(handler)
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/websocket.py", line 964, in _accept_connection
        await self._receive_frame_loop()
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/websocket.py", line 1118, in _receive_frame_loop
        await self._receive_frame()
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/websocket.py", line 1130, in _receive_frame
        data = await self._read_bytes(2)
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/websocket.py", line 1124, in _read_bytes
        data = await self.stream.read_bytes(n)
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/iostream.py", line 436, in read_bytes
        future = self._start_read()
      File "/opt/hostedtoolcache/Python/3.6.5/x64/lib/python3.6/site-packages/tornado/iostream.py", line 797, in _start_read
        assert self._read_future is None, "Already reading"
    AssertionError: Already reading
```

**Comment by jasongrout** [comment]
I just rebased on current master to again to make sure we are reflecting the very latest state of master.

**Comment by jasongrout** [comment]
To narrow things down a bit, I added skips and restructured tests so that tests pretty consistently pass locally. Let's see how the CI fares.

File: tests/test-application/src/shell.spec.ts [hunk 1]
```diff
@@ -24,6 +24,12 @@ class ContentWidget extends Widget {
 describe('LabShell', () => {
   let shell: LabShell;
 
+  beforeAll(() => {
```
Comment (afshin) [inline tests/test-application/src/shell.spec.ts]: Nice 👍

File: tests/test-apputils/src/dialog.spec.tsx [hunk 2]
```diff
@@ -53,7 +53,7 @@ class ValueWidget extends Widget {
 }
 
 describe('@jupyterlab/apputils', () => {
-  describe('Dialog', () => {
+  describe.skip('Dialog', () => {
```
Comment (afshin) [inline tests/test-apputils/src/dialog.spec.tsx]: All dialog tests needed to be skipped?

Comment (jasongrout) [reply tests/test-apputils/src/dialog.spec.tsx]: I was having lots of issues with various ones, so I was trying to get down to the subset of tests I needed to concentrate on. Skipping all was probably too broad, but it's still a WIP.

**Comment by Zsailer** [comment]
I'm diving into the server side of this issue today. I'll let y'all know if I find any new information.

**Comment by ian-r-rose** [comment]
Closing in favor of #6191

**Comment by ian-r-rose** [comment]
@Zsailer Were you able to reproduce the `'Already reading'` issue referenced here and in #6131?

**Comment by Zsailer** [comment]
Yeah, I can reproduce the `'Already reading'` error here, but I haven't tracked down the source. I'll keep working on it though...

**Pull request closed without merging**